### PR TITLE
Robustness test flake fix

### PR DESCRIPTION
**Pull request by sheidkamp**
# Description
Fix the cleanup of the gateway kube2e tests that use a validatingAdmissionWebhook Failure Policy of `Fail` by restoring the Failure Policy to what it was at the start of the test block.

## Testing steps
To validate the changes work as intended, run the gateway kube2e tests locally and confirm that at the end of the tests, the ValidatingWebhookConfiguration has `failurePolicy: Ignore`

The test that fails is this test:
```
	It("works, even with deleted services", func() {
		// Delete the k8s service behind the second echo app
		err = resourceClientset.KubeClients().CoreV1().Services(appService2.Namespace).Delete(ctx, appService2.Name, metav1.DeleteOptions{})
		Expect(err).NotTo(HaveOccurred())

		assertCanRouteSvc1NotSvc2()

		// roll pods to ensure we are resilient to pod restarts
		By("bounce gloo and envoy")
		scaleDeploymentTo(resourceClientset.KubeClients(), glooDeployment, 0)
		scaleDeploymentTo(resourceClientset.KubeClients(), envoyDeployment, 0)
		scaleDeploymentTo(resourceClientset.KubeClients(), glooDeployment, 1)
		scaleDeploymentTo(resourceClientset.KubeClients(), envoyDeployment, 1)

		assertCanRouteSvc1NotSvc2()
	})
```
and it fails when `DeleteSnapshot` is called in the `JustAfterEach` with an error message of:
```
failed calling webhook "gloo.gloo-system.svc": failed to call webhook: Post "[https://gloo.gloo-system.svc:443/validation?timeout=10s](https://gloo.gloo-system.svc/validation?timeout=10s)": dial tcp 10.96.175.96:443: connect: connection refused
```

I have not been able to replicate the flake itself, but can replicate the error message with the following steps:

- In with a gloo OSS cluster with a ValidatingWebhookConfiguration with a `failurePolicy: Fail`
- Create a secret:
```
kubectl create secret generic db-user-pass --namespace gloo-system \
  --from-literal=username=admin \
  --from-literal=password='S!B\*d$zDsb='
```
- Scale down the `gloo` deployment as in the test:
```
k scale --replicas=0 deployment/gloo -n gloo-system
```
- Attempt to delete the secret:
```
k delete secret db-user-pass -n gloo-system
```
and get an error similar to the one in the flake:
```
Error from server (InternalError): Internal error occurred: failed calling webhook "gloo.gloo-system.svc": failed to call webhook: Post "https://gloo.gloo-system.svc:443/validation?timeout=10s": dial tcp 10.96.245.94:443: connect: connection refused
```
- Scale the deployment up and immediately try to delete the secret
```
k scale --replicas=1 deployment/gloo -n gloo-system
k delete secret db-user-pass -n gloo-system
```
The deletion should fail and continue to fail for a few seconds.
- Recreate the secret if it was deleted:
```
kubectl create secret generic db-user-pass --namespace gloo-system \
  --from-literal=username=admin \
  --from-literal=password='S!B\*d$zDsb='
```
- Scale the deployment back down:
```
k scale --replicas=0 deployment/gloo -n gloo-system
```
- Update the `failurePolicy` to `Ignore`, as in the updated test cleanup
- Attempt to delete the secret:
```
k delete secret db-user-pass -n gloo-system
```
- Success!
```
secret "db-user-pass" deleted
```

These steps show that when a deployment is scaled down and then scaled back up as in the test it will cause the error we see for some amount of time if the `failurePolicy` is `Fail`. They also show that if the `failurePolicy` is `Ignore` as expected, the secret can be deleted even if there are no pods available for validation. This provides a likely explanation for the errors and proof of the fix.

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by bewebi — COMMENTED** [review]
Looks good

As a nit, I wonder if it makes sense to try and make the Get and Update functions more symmetrical since we get the whole webhook but only update failure policy

Alternatively we could have `GetFailurePolicy()` and `UpdateFailurePolicy()` or `GetValidatingWebhook()` and `PatchValidatingWebhook()`

I don't feel strongly about this, but wanted to be sure it was considered

**Comment by sheidkamp** [comment]
/kick https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=e37ea7a9-e4a4-4809-8287-1f24b315ebcb check for flake?

File: test/kube2e/util.go [hunk 1]
```diff
@@ -230,12 +230,25 @@ func GetTestHelper(ctx context.Context, namespace string) (*helper.SoloTestHelpe
...
 
-	_, err = kubeClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().Update(ctx, cfg, metav1.UpdateOptions{})
+	_, err := kubeClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().Update(ctx, cfg, metav1.UpdateOptions{})
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+}
+
+func GetValidatingWebhook(ctx context.Context, webhookName string) *admissionregv1.ValidatingWebhookConfiguration {
+	kubeClient := clienthelpers.MustKubeClient()
+	cfg, err := kubeClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(ctx, webhookName, metav1.GetOptions{})
 	ExpectWithOffset(1, err).NotTo(HaveOccurred())
```
Comment (davidjumani) [inline test/kube2e/util.go:252]: Offset = 2 since this is called within a get / set failure policy ?

**Review by davidjumani — DISMISSED** [review]

File: test/kube2e/util.go (hunk 1 above)
Comment (sheidkamp) [reply test/kube2e/util.go]: Good call, updated to split out a "base" `GetValidatingWebhook` and a `GetValidatingWebhookWithOffset`

**Comment by sheidkamp** [comment]
> Looks good
> 
> As a nit, I wonder if it makes sense to try and make the Get and Update functions more symmetrical since we get the whole webhook but only update failure policy
> 
> Alternatively we could have `GetFailurePolicy()` and `UpdateFailurePolicy()` or `GetValidatingWebhook()` and `PatchValidatingWebhook()`
> 
> I don't feel strongly about this, but wanted to be sure it was considered

I have used the `GetFailurePolicy()` and `UpdateFailurePolicy()` approach, with `GetValidatingWebhook()` as a helper.

If we find a second use case we can create `PatchValidatingWebhook()`

**Review by bewebi — APPROVED** [review]

**Comment by sheidkamp** [comment]
dozer?

**Pull request merged**